src/Pact/Native/Keysets.hs

```diff
@@ -82,6 +89,43 @@ defineKeyset g0 fi as = case as of
             runSysOnly $ enforceKeySet i (Just ksn) oldKs
             writeRow i Write KeySets ksn ks & success "Keyset defined"
 
+descKeySet :: GasRNativeFun e
+descKeySet g i [TLitString t] = do
+  r <- readRow (_faInfo i) KeySets (KeySetName t)
+  case r of
+    Just v -> computeGas' g i (GPostRead (ReadKeySet (KeySetName t) v)) $
+              return $ toTerm v
+    Nothing -> evalError' i $ "Keyset not found: " <> pretty t
+descKeySet _ i as = argsError i as
+
+descGuard :: GasRNativeFun e
+descGuard gas i [TGuard g _] = do
+  case g of
+    GKeySet ks ->
+      computeGas' gas i (GUserApp Defun) $
+      return $ toTObject TyAny def
+        [ ("type", tStr "KeySet")
+        , ("keyset", toTerm ks)
+        ]
+    GKeySetRef (KeySetName t) -> do
+      r <- readRow (_faInfo i) KeySets (KeySetName t)
+      case r of
+        Just v -> computeGas' gas i (GPostRead (ReadKeySet (KeySetName t) v)) $
+                  return $ toTObject TyAny def
+                    [ ("type", tStr "KeySetRef")
+                    , ("keyset", toTerm v)
+                    ]
+        Nothing -> evalError' i $ "Keyset not found: " <> pretty t
+
+      --descKeySet gas i [TLiteral (LString t) info]
+    GPact _ -> computeGas' gas i (GUserApp Defun) $
+      return $ toTObject TyAny def [ ("type", tStr "Pact") ]
+    GModule _ -> computeGas' gas i (GUserApp Defun) $
+      return $ toTObject TyAny def [ ("type", tStr "Module") ]
+    GUser _ -> computeGas' gas i (GUserApp Defun) $
+      return $ toTObject TyAny def [ ("type", tStr "User") ]
```

Comment:
OK this is where it gets interesting.

First, it should have `fun` and `args` from `UserGuard`. 

(Note that all of these (including other types) should probably try to stay consistent with JSON field names. This is maybe why a `details` field could be good, to make it really clear that we're enumerating the constructor JSON-style)

Second: wasn't it a motivation to show all keysets? This is where we might be able to traverse the guard structure. Let's look at the JSON of a gas station with a key:

```
{ "fun": "util.guards1.enforce-guard-any"
  "args": [
    [ { "pred": "keys-all",
        "keys": [ "9a4849687cbcfeb1f7c6510539638da576289508aedcc75f4d6ad3ed2623635c" ]
      },
      { "fun": "util.guards1.enforce-guard-all",
        "args": [
          [ { "fun": "coin.gas-only",
              "args": []
            },
            { "fun": "util.guards1.enforce-below-or-at-gas-price",
              "args": [ 1e-08 ]
            },
            { "fun": "util.guards1.enforce-below-or-at-gas-limit",
              "args": [ { "int": 10000 } ]
            }
          ] ]
      }
    ] ]
}
```
So, first, it might be nice to try to pretty-print in a `desc` field. More importantly, we can walk the args in that any time we find a `TGuard` argument, we can find keysets! Then we dump them in a `keysets` field. What do you think?

Reply:
This sounds promising!  I'm also thinking about the scenario where maybe you have a smart wallet that allows different keysets in different situations.  Is there a way the smart wallet module could somehow expose what keys it's accepting?  I'm imagining situations where it's not necessarily static.  For instance, maybe you have a smart wallet designed to make funds available to beneficiaries on the event of death or loss of owner's keys.  In this situation you might have a time delay.  If it's been less than a year since the last transaction, only allow the owner's key.  Otherwise allow some multi-sig of social recovery with maybe a family member plus an attorney or something.